budgetportal/tests/test_provincial_infrastructure_projects.py

```diff
@@ -168,3 +174,141 @@ def test_other_parties(self):
+        self.sources = [
+            "Equitable Share",
+            "Education Infrastructure Grant",
+            "Community Library Service Grant",
+        ]
+        for i in range(30):
+            ProvInfraProject.objects.create(
+                financial_year=self.fin_year,
+                IRM_project_id=i,
+                name="Project {}".format(i),
+                department="Department {}".format(i),
+                local_municipality="Local {}".format(i),
+                district_municipality="District {}".format(i),
+                province=random.choice(self.provinces),
+                status=random.choice(self.statuses),
+                primary_funding_source=random.choice(self.sources),
+                main_contractor="Contractor {}".format(i),
+                principle_agent="Principle Agent {}".format(i),
+                program_implementing_agent="Program Agent {}".format(i),
+                other_parties="Service Provider: XXXX{}".format(i),
+            )
+
+    def test_projects_per_page(self):
+        response = self.client.get(self.url, format="json")
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        number_of_results = len(response.data["results"])
+        self.assertLessEqual(number_of_results, 20)
+
+    def test_filter_by_department(self):
+        project = ProvInfraProject.objects.get(department=self.department)
+
+        data = {"department": self.department}
+        response = self.client.get(self.url, data)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        number_of_projects = len(response.data["results"])
+        response_data = response.data["results"][0]
+        self.assertEqual(number_of_projects, 1)
+        self.assertEqual(response_data["name"], project.name)
+        self.assertEqual(response_data["province"], project.province)
+        self.assertEqual(
+            response_data["local_municipality"], project.local_municipality
+        )
+        self.assertEqual(
+            response_data["district_municipality"], project.district_municipality
+        )
+
+    def test_filter_by_province(self):
+        projects = ProvInfraProject.objects.filter(province=self.province)
+
+        data = {"province": self.province}
+        response = self.client.get(self.url, data)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        number_of_projects = len(response.data["results"])
+        self.assertEqual(number_of_projects, projects.count())
+
+    def test_filter_by_status(self):
+        projects = ProvInfraProject.objects.filter(status=self.status)
+
+        data = {"status": self.status}
+        response = self.client.get(self.url, data)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        number_of_projects = len(response.data["results"])
+        self.assertEqual(number_of_projects, projects.count())
```

Comment:
rather do assertEqual(number_of_projects, 5) or whatever

It's a nice idea, but it fails when you accidentally change your data and you don't realise this test depended on it, and you end up with zero matches and your test doesn't mean anything any more.

The lesson here is test code should **not** be clever code. It should be blatantly obvious code and it's better for it to fail when refactoring than to accidentally pass despite being wrong.

Reply:
project.count() changed with hardcoded number.

https://github.com/vulekamali/datamanager/pull/190/commits/209ccd71587768f081ef8e2e4c8e7a798eeddb7f